tests/unittests/Foundation/NSDecimalNumberTests.mm

```diff
@@ -0,0 +1,330 @@
+    EXPECT_EQ(5, result._length);
+    EXPECT_EQ(0, result._isNegative);
+    EXPECT_EQ(-18, result._exponent);
+
+    //
+    testSubtract([[[NSDecimalNumber alloc] initWithDouble:10.5] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:10] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:0.5] autorelease]);
+
+    testSubtract([[[NSDecimalNumber alloc] initWithDouble:10.5] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:1000] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:-989.5] autorelease]);
+
+    testSubtract([[[NSDecimalNumber alloc] initWithDouble:-10.5] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:10] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:-20.5] autorelease]);
+
+    testSubtract([[[NSDecimalNumber alloc] initWithDouble:10.5] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:-1e3] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:1010.5] autorelease]);
+
+    NSDecimalNumber* d = [[[NSDecimalNumber alloc] initWithDouble:1012345678912345e-120] autorelease];
+    testSubtract(d, d, [NSDecimalNumber zero]);
+
+    testSubtract([[[NSDecimalNumber alloc] initWithInteger:2] autorelease], [NSDecimalNumber one], [NSDecimalNumber one]);
+}
+
+TEST(NSDecimalNumber, addition) {
+    // Note: Subtraction tests addition, since NSDecimalSubtract gets converted to NSDecimalAdd
+    unsigned short mantissa[NSDecimalMaxSize] = { 5272, 10, 0, 0, 0, 0, 0, 0 };
+    NSDecimalNumber* dNumber = [[[NSDecimalNumber alloc] initWithLongLong:6560451.0] autorelease];
+    NSDecimal decimal1 = [dNumber decimalValue];
+
+    dNumber = [[[NSDecimalNumber alloc] initWithDouble:45869] autorelease];
+    double maxx = DBL_MAX;
+    NSDecimal decimal2 = [dNumber decimalValue];
+
+    NSDecimal result = { 0 };
+
+    NSDecimalAdd(&result, &decimal1, &decimal2, 0);
+
+    ASSERT_EQ(2, result._length);
+    ASSERT_EQ(1, result._exponent);
+    ASSERT_EQ(0, result._isNegative);
+
+    EXPECT_ARRAY_EQUAL(mantissa, result._mantissa);
+
+    testAddition([[[NSDecimalNumber alloc] initWithDouble:19605] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:6536] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:26141] autorelease]);
+
+    testAddition([[[NSDecimalNumber alloc] initWithDouble:10.5] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:-1e3] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:-989.5] autorelease]);
+}
+
+TEST(NSDecimalNumber, NSDecimalCompare) {
+    NSDecimalNumber* dNumber = [[[NSDecimalNumber alloc] initWithUnsignedLongLong:9223372534876] autorelease];
+    NSDecimal right = [dNumber decimalValue];
+
+    NSDecimal left = { 0 };
+    unsigned long long value = 10;
+    memcpy(&left._mantissa, &value, NSDecimalMaxSize * sizeof(short));
```

Comment:
since we are limiting this by length, we'll ignore the arbitrary values.

---
In reply to: [80568168](https://github.com/Microsoft/WinObjC/pull/1032#discussion_r80568168) [](ancestors = 80568168)